swift/ql/lib/codeql/swift/dataflow/internal/FlowSummaryImplSpecific.qll

```diff
@@ -32,29 +32,63 @@ ArgumentPosition callbackSelfParameterPosition() { result instanceof ThisArgumen
 SummaryCall summaryDataFlowCall(SummaryNode receiver) { receiver = result.getReceiver() }
 
 /** Gets the type of content `c`. */
-DataFlowType getContentType(ContentSet c) { any() }
+DataFlowType getContentType(ContentSet c) {
+  exists(Content cc |
+    c.isSingleton(cc) and
+    result = cc.getType()
+  )
+}
 
 /** Gets the type of the parameter at the given position. */
-DataFlowType getParameterType(SummarizedCallable c, ParameterPosition pos) { any() }
+DataFlowType getParameterType(SummarizedCallable c, ParameterPosition pos) {
+  result = getDataFlowType(c.getParam(pos.(PositionalParameterPosition).getIndex()).getType())
+  or
+  pos instanceof ThisParameterPosition and
+  result = getDataFlowType(c.getSelfParam().getType())
+}
 
 /** Gets the return type of kind `rk` for callable `c`. */
 bindingset[c]
-DataFlowType getReturnType(SummarizedCallable c, ReturnKind rk) { any() }
+DataFlowType getReturnType(SummarizedCallable c, ReturnKind rk) {
+  rk instanceof NormalReturnKind and
+  result = getDataFlowType(c.getResultType())
+  or
+  exists(ParamDecl p |
+    p = c.getParam(rk.(ParamReturnKind).getIndex()) and
+//    p.isInout() and
+    result = getDataFlowType(p.getType())
+  )
+  or
+  rk.(ParamReturnKind).getIndex() = -1 and
+//  c.getSelfParam().isInout() and
+  result = getDataFlowType(c.getSelfParam().getType())
+}
 
 /**
  * Gets the type of the parameter matching arguments at position `pos` in a
  * synthesized call that targets a callback of type `t`.
  */
-DataFlowType getCallbackParameterType(DataFlowType t, ArgumentPosition pos) { any() }
+DataFlowType getCallbackParameterType(DataFlowType t, ArgumentPosition pos) {
+  exists(t) and 
+  exists(pos) and
+  result.asType() instanceof AnyType
+}
```

Comment:
What's going on here?

Reply:
This predicate only applies to nodes that are synthesized from a callback in a modeled function. The `AnyType` is the top of Swift's type hierarchy (although it's not included in the `getABaseType` relationship), so we're allowing any type here.

Reply:
This looks like a cartesian product between all `DataFlowType`s and all `ArgumentPosition`s, right? This was fine before because `DataFlowType` was a unit type, but this looks like a real CP now, though? Or is the assumption that this fine because there aren't _that_ many `ArgumentPosition`s?